src/core/utils/fileutils.cpp

```diff
@@ -318,3 +321,250 @@ void FileUtils::addImageStamp( const QString &imagePath, const QString &text )
+    return false;
+  }
+
+  QFile file( filePath );
+  QFileInfo fileInfo( filePath );
+  QDir directory = fileInfo.dir();
+
+  bool isLikelyWritable = false;
+
+#ifdef Q_OS_ANDROID
+  QString appStorage = QStandardPaths::writableLocation( QStandardPaths::AppDataLocation );
+  isLikelyWritable = filePath.startsWith( appStorage );
+#elif defined( Q_OS_IOS )
+  QString appStorage = QStandardPaths::writableLocation( QStandardPaths::AppDataLocation );
+  isLikelyWritable = filePath.startsWith( appStorage );
+#else
+  QFileInfo dirInfo( directory.absolutePath() );
+  isLikelyWritable = dirInfo.isWritable();
+#endif
+
+  if ( !isLikelyWritable )
+  {
+    qWarning() << QStringLiteral( "Writing to %1 may fail due to platform restrictions. Use PlatformUtilities.applicationDirectory() for a safe location." ).arg( filePath );
+  }
+
+  if ( !directory.exists() )
+  {
+    if ( !directory.mkpath( "." ) )
+    {
+      qDebug() << QStringLiteral( "Failed to create directory for file: %1. This may be due to permission restrictions." ).arg( filePath );
+      return false;
+    }
+  }
+
+  if ( file.open( QIODevice::WriteOnly ) )
+  {
+    qint64 bytesWritten = file.write( content );
+    file.close();
+
+    if ( bytesWritten != content.size() )
+    {
+      qDebug() << QStringLiteral( "Failed to write all data to file: %1" ).arg( filePath );
+      return false;
+    }
+
+    return true;
+  }
+  else
+  {
+    QString errorMsg = file.errorString();
+    if ( file.error() == QFile::PermissionsError )
+    {
+      errorMsg += QStringLiteral( " - This may be due to platform security restrictions." );
+    }
+    qDebug() << QStringLiteral( "Failed to open file for writing: %1" ).arg( errorMsg );
+    return false;
+  }
+}
+
+QVariantMap FileUtils::getFileInfo( const QString &filePath )
+{
+  QVariantMap info;
+
+  if ( !isWithinProjectDirectory( filePath ) )
+  {
+    qWarning() << QStringLiteral( "Security warning: Attempted to access file info outside project directory: %1" ).arg( filePath );
+    info["exists"] = false;
+    info["error"] = QStringLiteral( "Access denied: File is outside the current project directory" );
+    return info;
+  }
+
+  QFileInfo fileInfo( filePath );
+  QFile file( filePath );
+
+  if ( fileInfo.exists() )
+  {
+    info["exists"] = true;
+    info["fileName"] = fileInfo.fileName();
+    info["filePath"] = fileInfo.absoluteFilePath();
+    info["fileSize"] = fileInfo.size();
+    info["lastModified"] = fileInfo.lastModified();
+    info["suffix"] = fileInfo.suffix();
+
+    QMimeDatabase db;
+    info["mimeType"] = db.mimeTypeForFile( filePath ).name();
+
+    QByteArray md5Hash = fileChecksum( filePath, QCryptographicHash::Md5 );
+    if ( !md5Hash.isEmpty() )
+    {
+      info["md5"] = md5Hash.toHex();
+    }
+    else
+    {
+      info["md5Error"] = QStringLiteral( "Could not calculate MD5 hash - possibly due to permission restrictions" );
```

Comment:
Same as above. I am wondering if we should insure all keys always have associated values.

Reply:
Will respect it in the following commit. Will take a moment as I first check if tests pass before pushing to upstream. Thanks @nirvn !